### PR TITLE
DDP Support Refactor and Cleanup

### PR DESCRIPTION
**Pull request by ethankan**
- Refactored the Key-Value cache mechanism into `kvCache`, which coordinates the "sync" from cloud, and the update to the cloud.
- Allow key-value read-only access even when the session is in recovery state.
- Some changes to the callback definitions.
- Added godoc for exported functions/variables.
- Finished the example program.

### CODE REVIEW

File: core.go [hunk 1]
```diff
@@ -52,15 +52,17 @@ var (
 	infoLogger  = log.New(os.Stdout, "[MODE - INFO] ", log.LstdFlags)
 	errorLogger = log.New(os.Stderr, "[MODE - ERROR] ", log.LstdFlags)
 
-	maxDeviceEventAttempts      uint = 3
-	maxDeviceKeyValueAttempts   uint = 3
-	deviceEventRetryInterval         = time.Second * 5
-	deviceKeyValueRetryInterval      = time.Second * 5
+	maxDeviceEventAttempts   uint = 3
```
Comment (honteng) [inline core.go]: why not aligned?

File: examples/ddp/main.go [hunk 2]
```diff
@@ -9,22 +13,42 @@ import (
...
+		}
+	})
+
+	mode.SetKeyValueStoredCallback(func(_ *mode.DeviceContext, kv *mode.KeyValue) {
+		// If the key 'msg' is updated, echo it back to the key 'echo'.
+		if kv.Key == "msg" {
+			fmt.Printf("Echoing msg %v\n", kv.Value)
+
+			if err := mode.SetKeyValue("echo", kv.Value); err != nil {
+				fmt.Printf("Failed to set key-value: %s\n", err.Error())
```
Comment (honteng) [inline examples/ddp/main.go]: %v", err) is simpler

**Comment by ethankan** [comment]
I'm also considering removing the option to use websocket, since adding DDP support in websocket will require more refactoring, and it would be odd if the DDP feature is not available for websocket. May as well just remove the websocket stuff.

It would be a breaking change,  will need to be released as v2.0.0.

File: kv_cache.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+		logError("[keyValueCache] received sync message with unknown action '%s'", kvSync.Action)
+	}
+}
+
+func (cache *keyValueCache) getKeyValue(key string) (*KeyValue, error) {
+	resErr := make(chan error)
+	resKV := make(chan *KeyValue)
+
+	cache.access <- func() {
+		if !cache.initialized {
```
Comment (honteng) [inline kv_cache.go:206]: ah nice design pattern!

File: mqtt.go [hunk 4]
```diff
@@ -199,18 +199,17 @@ func (mc *mqttConn) handleCommandMsg(p *packet.PublishPacket) error {
...
 func (mc *mqttConn) handleKeyValueMsg(p *packet.PublishPacket) error {
-	var kv ActionKeyValue
-	if err := decodeOpaqueJSON(p.Message.Payload, &kv); err != nil {
-		return fmt.Errorf("message data is not valid key value JSON: %s", err.Error())
-	}
+	var kvSync keyValueSync
 
-	kv.Payload = p.Message.Payload
+	if err := decodeOpaqueJSON(p.Message.Payload, &kvSync); err != nil {
+		return fmt.Errorf("message data is not valid key-value sync JSON: %s", err.Error())
```
Comment (honteng) [inline mqtt.go:205]: ditto you can use %v

File: mqtt.go [hunk 5]
```diff
@@ -313,27 +312,28 @@ func (mc *mqttConn) runPinger() {
...
-				logError("[MQTT] failed to send event: %s", err.Error())
+				logError("[MQTT] publisher failed to send event: %s", err.Error())
 			}
-		case kv := <-mc.sendKeyValue:
-			if err := mc.setKeyValue(kv); err != nil {
-				logError("[MQTT] failed to set keyValue: %s", err.Error())
+
+		case kvSync := <-mc.kvPush:
+			if err := mc.sendKeyValueUpdate(kvSync); err != nil {
+				logError("[MQTT] publisher failed to send key-value update: %s", err.Error())
```
Comment (honteng) [inline mqtt.go:336]: ditto

File: session.go [hunk 6]
```diff
@@ -1,7 +1,7 @@
 package mode
 
 import (
-	"encoding/json"
+	//	"encoding/json"
```
Comment (honteng) [inline session.go]: Delete please if not used

File: session.go [hunk 7]
```diff
@@ -154,16 +157,31 @@ func SetSessionStateCallback(f SessionStateCallback) {
...
 }
 
-func SetKvDeleteHandler(h KvDeleteHandler) {
-	kvDeleteHandler = h
+// SetKeyValueDeletedCallback designates a function to be called whenever a
+// key-value pair has been deleted by someone else.
+//
+// IMPORTANT: This callback function runs in a separate goroutine. Please take
+// care to handle any concurrency issues.
+func SetKeyValueDeletedCallback(f KeyValueDeletedCallback) {
```
Comment (honteng) [inline session.go]: what if delete and set key value happens almost at the same time?

The callback may be called out of order?

For example, if DELETE -> SET "key" happens, then the order will be guaranteed?

**Comment by honteng** [comment]
fine with deleting websocket.

VRC is using MQTT only so it should be fine.

**Pull request merged**